test/extended/util/configv1shim_test.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	err = client.ConfigV1().Infrastructures().DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set(map[string]string{"deleteLabel": "somevalue2"})).String(),
+	})
+	if err != nil {
+		t.Fatalf("Expected no error for a DeleteCollection request, got %q instead", err)
+	}
+}
+
+func TestConfigClientShimWatchRequest(t *testing.T) {
+	tests := []struct {
+		name                string
+		staticObjects       []runtime.Object
+		realObjects         []runtime.Object
+		expectedWatchEvents []watch.Event
+	}{
+		{
+			name: "merging static and real objects, not object override",
+			staticObjects: []runtime.Object{
+				createInfrastructureObject("cluster"),
+			},
+			realObjects: []runtime.Object{
+				createInfrastructureObject("cluster2"),
+				createInfrastructureObject("cluster3"),
+			},
+			expectedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: createInfrastructureObject("cluster")},
+				{Type: watch.Added, Object: createInfrastructureObject("cluster2")},
+				{Type: watch.Added, Object: createInfrastructureObject("cluster3")},
+			},
+		},
+		{
+			name: "merging static and real objects, static object override",
+			staticObjects: []runtime.Object{
+				createInfrastructureObject("cluster"),
+			},
+			realObjects: []runtime.Object{
+				createInfrastructureObject("cluster"),
+				createInfrastructureObject("cluster2"),
+			},
+			expectedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: createInfrastructureObject("cluster")},
+				{Type: watch.Added, Object: createInfrastructureObject("cluster2")},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			configClient := fakeconfigv1client.NewSimpleClientset()
+
+			client := NewConfigClientShim(
+				configClient,
+				test.staticObjects,
+			)
+
+			// The watch request has to created first when using the fake clientset
+			resultChan, err := client.ConfigV1().Infrastructures().Watch(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				t.Fatalf("Expected no error, got %q instead", err)
+			}
+			defer resultChan.Stop()
+
+			// And only then the fake clientset can be populated to generate the watch event
+			for _, obj := range test.realObjects {
+				configClient.Tracker().Add(obj)
+			}
+
+			// verify the watch events
+			ticker := time.NewTicker(500 * time.Millisecond)
+			size := len(test.expectedWatchEvents)
+			eventCounter := 0
+			for i := 0; i < size; i++ {
+				select {
+				case item := <-resultChan.ResultChan():
+					diff := cmp.Diff(test.expectedWatchEvents[i], item)
+					if diff != "" {
+						t.Errorf("test '%s' failed. Results are not deep equal. mismatch (-want +got):\n%s", test.name, diff)
+					}
+					eventCounter++
+				case <-ticker.C:
+					t.Errorf("failed waiting for watch event")
+				}
+			}
+			if eventCounter < size {
```

Comment:
nit: this could also test if we got more events than expectedWatchEvents

nice fake watcher though

Reply:
Done